pkg/storage/gcp/saver.go

```diff
@@ -0,0 +1,101 @@
+package gcp
+
+import (
+	"fmt"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/bketelsen/buffet"
+	"github.com/gobuffalo/buffalo"
+	"google.golang.org/api/option"
+)
+
+// Storage implements the (github.com/gomods/pkg/storage).Saver interface
+type Storage struct {
+	bucket *storage.BucketHandle
+}
+
+// New returns a new Storage instance
+// authenticated using the provided ClientOptions
+func New(ctx buffalo.Context, cred option.ClientOption) (*Storage, error) {
+	client, err := storage.NewClient(ctx, cred)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new client: %s", err)
+	}
+	// The bucket MUST already exist
+	bkt := client.Bucket("gomodules")
+	return &Storage{bucket: bkt}, nil
+}
+
+// Save uploads the modules .mod, .zip and .info files for a given version
+func (s *Storage) Save(ctx buffalo.Context, module, version string, mod, zip, info []byte) error {
+	sp := buffet.ChildSpan("storage.save", ctx)
+	defer sp.Finish()
+
+	errs := make(chan error, 3)
+	// dispatch go routine for each file to upload
+	// TODO: make this compact
+	go func(errs chan<- error) {
+		sp := buffet.ChildSpan("storage.save.module", ctx)
+		defer sp.Finish()
+
+		if err := save(ctx, s.bucket, fmt.Sprintf("%s/@v/%s.mod", module, version), mod); err != nil {
+			errs <- err
+		} else {
+			errs <- nil
+		}
+	}(errs)
+	go func(errs chan<- error) {
+		sp := buffet.ChildSpan("storage.save.zip", ctx)
+		defer sp.Finish()
+
+		if err := save(ctx, s.bucket, fmt.Sprintf("%s/@v/%s.zip", module, version), zip); err != nil {
```

Comment:
we need to abstract away the creation of blob name, you can use paths package aaron added recently.

Reply:
I had a quick look at `pkg/paths` and am not sure how to use it for creating these blob names.

Reply:
Not sure either. @michalpristas maybe we should add something? Could use it [here](https://github.com/gomods/athens/blob/2d25ccb3dea2bab4e208849cae58039ddf26e14e/pkg/storage/azurecdn/storage.go#L51) too

Reply:
@arschles sure thats' what i meant. we compose those filenames on too many places, let's ,make single func which handles this names, this will help with possible refactoring later